File: frontend/src/lib/api/season.ts
import { authFetch } from '../auth'
import { readApiError } from './http'

export type SeasonSummary = {
  id: number
  name: string
  status: string
  startedAt: string
  closedAt?: string | null
  readyToClose: boolean
}

export type Season = SeasonSummary & {
  closingFantasyLeagueId?: number | null
  closingFantasyLeagueName?: string | null
}

export type ClosePreviewTournament = {
  id: number
  link: string
  name: string | null
  startDate: string | null
  endDate: string | null
  finished: boolean
  selected: boolean
  isFantasySource: boolean
}

export type SeasonClosePreview = {
  season: Season
  tournaments: ClosePreviewTournament[]
  closingFantasyLeagueId?: number | null
}

export async function fetchCurrentSeason(): Promise<Season | null> {
  const res = await authFetch('/api/seasons/current')
  if (res.status === 404) return null
  if (!res.ok) throw new Error(await readApiError(res, `season uplink failed (${res.status})`))
  const data = (await res.json()) as { season: Season }
  return data.season ?? null
}

export async function fetchSeasonClosePreview(): Promise<SeasonClosePreview> {
  const res = await authFetch('/api/seasons/close-preview')
  if (!res.ok) throw new Error(await readApiError(res, `season preview failed (${res.status})`))
  return (await res.json()) as SeasonClosePreview
}

export async function closeSeason(tournamentIds: number[]): Promise<Season> {
  const res = await authFetch('/api/seasons/close', {
    method: 'POST',
    body: JSON.stringify({ tournamentIds }),
  })
  if (!res.ok) throw new Error(await readApiError(res, `close season failed (${res.status})`))
  const data = (await res.json()) as { season: Season }
  return data.season
}

File: backend/internal/model/season.go
package model

// Season is one rating cycle.
type Season struct {
	ID                      int64   `json:"id"`
	Name                    string  `json:"name"`
	Status                  string  `json:"status"`
	StartedAt               string  `json:"startedAt"`
	ClosedAt                *string `json:"closedAt,omitempty"`
	ReadyToClose            bool    `json:"readyToClose"`
	ClosingFantasyLeagueID  *int64  `json:"closingFantasyLeagueId,omitempty"`
	ClosingFantasyLeagueName *string `json:"closingFantasyLeagueName,omitempty"`
}

// SeasonSummary is a compact season reference for list responses.
type SeasonSummary struct {
	ID           int64   `json:"id"`
	Name         string  `json:"name"`
	Status       string  `json:"status"`
	StartedAt    string  `json:"startedAt"`
	ClosedAt     *string `json:"closedAt,omitempty"`
	ReadyToClose bool    `json:"readyToClose"`
}

// ClosePreviewTournament is a tournament eligible for season-close rating selection.
type ClosePreviewTournament struct {
	ID           int64   `json:"id"`
	Link         string  `json:"link"`
	Name         *string `json:"name"`
	StartDate    *string `json:"startDate"`
	EndDate      *string `json:"endDate"`
	Finished     bool    `json:"finished"`
	Selected     bool    `json:"selected"`
	IsFantasySource bool `json:"isFantasySource"`
}

// SeasonClosePreview is the admin season-close screen payload.
type SeasonClosePreview struct {
	Season                 Season                   `json:"season"`
	Tournaments            []ClosePreviewTournament `json:"tournaments"`
	ClosingFantasyLeagueID *int64                   `json:"closingFantasyLeagueId,omitempty"`
}
